plugins/dashboard_plugin.py

```diff
@@ -96,53 +67,69 @@ def list(self):
+
         data = []
+
         for d in dags:
-            last_run = dag.get_last_dagrun(d.dag_id, session,include_externally_triggered=True)
+            last_run = dag.get_last_dagrun(d.dag_id,
+                                           self.AIRFLOW_SESSION,
+                                           include_externally_triggered=True)
 
             if last_run:
                 run_state = last_run.get_state()
+                run_date_info = self._get_localized_time(last_run.execution_date)
 
-                run_date = last_run.execution_date
+                last_successful, last_successful_info = self._get_last_successful_dagrun(d)
 
-                pst_run_time = run_date.astimezone(PACIFIC_TIMEZONE)
-                cst_run_time = run_date.astimezone(CENTRAL_TIMEZONE)
-                run_date_info = {
-                    'pst_time': datetime.strftime(pst_run_time, "%m/%d/%y %I:%M %p"),
-                    'cst_time': datetime.strftime(cst_run_time, "%m/%d/%y %I:%M %p")
-                }
+                next_scheduled = d.following_schedule(datetime.now(pytz.utc))
+                next_scheduled_info = self._get_localized_time(next_scheduled)
 
-                now = datetime.now(pytz.utc)
-                next_scheduled = d.following_schedule(now)
-
-                pst_next_scheduled = next_scheduled.astimezone(PACIFIC_TIMEZONE)
-                cst_next_scheduled = next_scheduled.astimezone(CENTRAL_TIMEZONE)
-                next_scheduled_info = {
-                    'pst_time': datetime.strftime(pst_next_scheduled, "%m/%d/%y %I:%M %p"),
-                    'cst_time': datetime.strftime(cst_next_scheduled, "%m/%d/%y %I:%M %p")
-                }
             else:
                 run_state = None
-                run_date_info = None
-                next_scheduled_info = None
+                run_date_info = {}
+
+                last_successful = None
+                last_successful_info = {}
+
+                next_scheduled_info = {}
 
             dag_info = {
                 'name': d.dag_id,
                 'description': d.description,
                 'run_state': run_state,
                 'run_date': run_date_info,
+                'last_successful': last_successful,
+                'last_successful_date': last_successful_info,
                 'next_scheduled': next_scheduled_info
             }
 
             data.append(dag_info)
 
         return data
 
+    def get_last_successful_dagrun(self, dags):
+        last = max(
+            (dag for dag in dags if dag['last_successful_date'].get('pst_time')),
+            key=lambda x: x['last_successful_date']['pst_time']
+        )
+
+        return last['last_successful'], last['last_successful_date']
+
+    def get_next_dagrun(self, dags):
+        return min(
+            (dag for dag in dags if dag['next_scheduled'].get('pst_time')),
+            key=lambda x: x['next_scheduled']['pst_time']
+        )
```

Comment:
Total nitpick, but all of the other `get_*_dagrun` methods return a tuple, while this returns a `DagRun` object. It would be helpful to standardize on one of the two APIs.

Reply:
I absolutely agree! The downside is that this will change the data the template expects and require further changes from the original implementation, and (believe it or not, haha) I wanted to minimize changes here as much as possible. Would it be ok with you if I opened a code quality issue for this, but didn't make the change in this PR?

Reply:
Actually, nvm, I'll go ahead and do this!